src/main/protobuf/bazel_output_service.proto

```diff
@@ -0,0 +1,324 @@
+message StageArtifactsRequest {
+  message Artifact {
+    // path is relative to StartBuildResponse.output_path.
+    string path = 1;
+    // Describe how to stage the artifact.
+    google.protobuf.Any locator = 2;
+  }
+
+  // The identifier of the build. Server uses this to determine which output
+  // path needs to be modified.
+  string build_id = 1;
+
+  repeated Artifact artifacts = 2;
+}
+
+message StageArtifactsResponse {
+  message Response {
+    // If the status has a code other than `OK`, it indicates that the artifact
+    // cannot be staged.
+    //
+    // Errors:
+    // * `NOT_FOUND`: The requested Artifact is not in the CAS.
+    google.rpc.Status status = 1;
+  }
+
+  // The response for each of the requested artifacts, using the same order as
+  // requested. This means that this list has the same length as
+  // StageArtifactsRequest.artifacts.
+  repeated Response responses = 1;
+}
+
+message FinalizeArtifactsRequest {
+  message Artifact {
+    // path is relative to StartBuildResponse.output_path.
+    string path = 1;
+    // Expected digest for this file. This allows server to detect if the file
+    // has been changed after Bazel finished creating the file and the
+    // corresponding FinalizeArtifactsRequest is processed. This field is
+    // optional for regular files, and should be omitted for directories and
+    // symlinks.
+    //
+    // The concrete type of the digest depending on the CAS Bazel connects to.
+    // In case of a REv2 CAS, the type is
+    // [Digest][build.bazel.remote.execution.v2.Digest].
+    google.protobuf.Any digest = 2;
+  }
+
+  // The identifier of the build. Server uses this to determine which output
+  // path needs to be modified.
+  string build_id = 1;
+
+  repeated Artifact artifacts = 2;
+}
+
+message FinalizeArtifactsResponse {
+  // Intentionally empty for future extensibility.
+}
+
+message FinalizeBuildRequest {
+  // The identifier of the build that should be finalized.
+  string build_id = 1;
+
+  // Whether the build completed successfully. The remote output service may,
+  // for example, use this option to apply different retention policies that
+  // take the outcome of the build into account.
+  bool build_successful = 2;
+}
+
+message FinalizeBuildResponse {
+  // Intentionally empty for future extensibility.
+}
+
+message BatchStatRequest {
+  // The identifier of the build. The remote output service uses this to
+  // determine which output path needs to be inspected.
+  string build_id = 1;
+
+  // Paths whose status needs to be obtained. The server MUST resolve the path
+  // using lstat semantic, i.e. components except the last one must be resolved
+  // if they are symlinks.
+  //
+  // Path is relative to StartBuildResponse.output_path.
+  repeated string paths = 2;
+}
+
+message BatchStatResponse {
+  message StatResponse {
+    // The status of the file. If the file corresponding with the requested path
+    // does not exist, this field will be null.
```

Comment:
"If no file exists at the requested path, this field MUST be unset."

I think the implication here is that an unresolvable symlink or a special file should still set `file_stat` to an empty message, because it's useful to distinguish "does not exist" from "exists, but the type is unsupported" (special file) or "could not determine whether it exists" (unresolvable symlink). This is a bit too easy to get subtly wrong.

Are there any situations where being able to distinguish these 3 cases is an advantage? An alternative would be to either provide one of the `oneof` fields, or leave the entire `BatchStatResponse` unset (meaning: may or may not exist, you have to stat the filesystem directly to find out). This would also let us inline the `oneof` directly into `StatResponse` and dispense with `FileStat`.

Reply:
Actually, I'm convinced that it's worthwhile to be able to distinguish nonexistence from the other two: in the case where all of the N files in a BatchStat are gone, it would be really bad to fall back to N filesystem stats. Please ignore the above, other than the suggested rewording,.

Reply:
Done.